nexus/src/db/datastore/silo_group.rs

```diff
@@ -0,0 +1,208 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! [`DataStore`] methods related to [`SiloGroup`]s.
+
+use super::DataStore;
+use crate::authz;
+use crate::context::OpContext;
+use crate::db;
+use crate::db::datastore::RunnableQuery;
+use crate::db::error::public_error_from_diesel_pool;
+use crate::db::error::ErrorHandler;
+use crate::db::error::TransactionError;
+use crate::db::lookup::LookupPath;
+use crate::db::model::SiloGroup;
+use crate::db::model::SiloGroupMembership;
+use async_bb8_diesel::AsyncRunQueryDsl;
+use async_bb8_diesel::{AsyncConnection, OptionalExtension};
+use chrono::Utc;
+use diesel::prelude::*;
+use omicron_common::api::external::CreateResult;
+use omicron_common::api::external::DeleteResult;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupResult;
+use omicron_common::api::external::UpdateResult;
+use uuid::Uuid;
+
+impl DataStore {
+    pub async fn silo_group_create_query(
+        opctx: &OpContext,
+        authz_silo: &authz::Silo,
+        silo_group: SiloGroup,
+    ) -> Result<impl RunnableQuery<SiloGroup>, Error> {
+        opctx.authorize(authz::Action::CreateChild, authz_silo).await?;
+
+        use db::schema::silo_group::dsl;
+        Ok(diesel::insert_into(dsl::silo_group)
+            .values(silo_group)
+            .returning(SiloGroup::as_returning()))
+    }
+
+    pub async fn silo_group_create(
+        &self,
+        opctx: &OpContext,
+        authz_silo: &authz::Silo,
+        silo_group: SiloGroup,
+    ) -> CreateResult<SiloGroup> {
+        DataStore::silo_group_create_query(opctx, authz_silo, silo_group)
+            .await?
```

Comment:
Based on the caller, I think in this case you want to `ON CONFLICT DO NOTHING` -- otherwise if two people log in at the same time with the same new group, one might fail spuriously.

I'm not sure if we should have a different naming convention for a function like that.  I sometimes use `ensure` as opposed to `create` to mean "ensure this exists, whether it did or not".

Reply:
Right, I initially thought because `silo_user_from_authenticated_subject` calls `silo_group_lookup_or_create_by_name` we wouldn't see this issue, but the whole login is not wrapped in a transaction. Changed in 9ee416f4.